File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/FeedbackLoopData.cs
using Newtonsoft.Json;

namespace Microsoft.Teams.AI.Application
{
    /// <summary>
    /// Data returned when the thumbsup or thumbsdown button is clicked and response is received.
    /// </summary>
    public class FeedbackLoopData
    {
        /// <summary>
        /// The action name.
        /// </summary>
        public string ActionName { get; set; } = "feedback";

        /// <summary>
        /// The action value.
        /// </summary>
        [JsonProperty(PropertyName = "actionValue")]
        public FeedbackLoopDataActionValue? ActionValue { get; set; }

        /// <summary>
        /// The activity ID that the feedback provided on.
        /// </summary>
        [JsonProperty(PropertyName = "replyToId")]
        public string? ReplyToId { get; set; }
    }

    /// <summary>
    /// The feedback loop data's action value.
    /// </summary>
    public class FeedbackLoopDataActionValue
    {
        /// <summary>
        /// Either "like" or "dislike"
        /// </summary>
        [JsonProperty(PropertyName = "reaction")]
        public string? Reaction { get; set; }

        /// <summary>
        /// The feedback provided by the user when prompted with "What did you lke/dislike?"
        /// </summary>
        [JsonProperty(PropertyName = "feedback")]
        public string? Feedback { get; set; }
    }
}

File: python/packages/ai/teams/feedback_loop_data.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, Literal, Union

from dataclasses_json import DataClassJsonMixin, config, dataclass_json


@dataclass_json
@dataclass
class FeedbackLoopData(DataClassJsonMixin):
    """
    Data returned when the thumbs up or down button
    is clicked and response sent when enable_feedback_loop
    is set to true in the AI Module.
    """

    action_value: FeedbackLoopActionValue = field(metadata=config(field_name="actionValue"))

    reply_to_id: str
    "The activity ID that the feedback was provided on."

    action_name: Literal["feedback"] = "feedback"


@dataclass_json
@dataclass
class FeedbackLoopActionValue(DataClassJsonMixin):
    """
    The value data provided with the feedback
    """

    reaction: Literal["like", "dislike"]
    "The reaction"

    feedback: Union[str, Dict[str, Any]]
    "The response the user provides after pressing one of the feedback buttons."
